Fix segfault in static_show, by using correct `vt` instead of `typeof(v)`

NHDaly opened the pull request:
This fixes a segfault introduced in #38049.

Co-Authored-By: @Sacha0

NHDaly committed:
Fix segfault in static_show, by using correct `vt` instead of `typeof(v)`

This fixes a segfault introduced in #38049.

Co-Authored-By: Sacha Verweij <sacha.verweij@alumni.stanford.edu>

vtjnash commented on src/rtutils.c:
```diff
@@ -999,7 +999,7 @@ static size_t jl_static_show_x_(JL_STREAM *out, jl_value_t *v, jl_datatype_t *vt
         n += jl_printf(out, ")");
     }
-    else if (jl_function_type && jl_isa(v, (jl_value_t*)jl_function_type)) {
+    else if (jl_function_type && jl_subtype(vt, (jl_value_t*)jl_function_type)) {
```
I wonder why this doesn't fail the analyzegc pass, as this function is prohibited from here

NHDaly replied:
😱 
Which function is prohibited? The old code, or the new code, or both?

vtjnash replied:
I'm not sure why `jl_is_operator` isn't failing the analyzegc pass either, as that's also prohibited by our annotations here

NHDaly replied:
It turns out that we made a mistake in the first PR, and hadn't noticed this comment:
https://github.com/JuliaLang/julia/blob/8f54d34c02206f6d8976e6e7916b51055a0f7a19/src/rtutils.c#L644-L648

This PR is attempting to now follow the rules of that comment.

Keno replied:
The analyzer sometimes bails on big functions. This one definitely qualifies

Keno replied:
But yeah, you're technically not allowed to use jl_subtype (or jl_isa). You may look at the `->super` field of the datatype though and see if you hit function eventually.

NHDaly replied:
Thanks @Keno. I believe I've fixed this in 2e5f22c1d7, by implementing a while loop using `->super`, as you suggest.

Could you take another look when you have a moment? Thanks! :)

KristofferC commented:
How often does this happen (aka how bad is it for PackageCompiler)? Also, can it be tested?

Sacha0 commented:
> How often does this happen (aka how bad is it for PackageCompiler)?

Every build on our end, regrettably.

NHDaly committed:
Fix illegal GC-violating call to `jl_subtype` in static_show

Implement custom iterative check via `->super`

NHDaly commented on src/rtutils.c:
```diff
@@ -641,6 +641,19 @@ static size_t jl_static_show_x_sym_escaped(JL_STREAM *out, jl_sym_t *name) JL_NO
+// explained in the comment on `jl_static_show_x_()`, below.
+// This function checks if `vt <: Function` without triggering GC.
+static int jl_static_is_function_(jl_datatype_t *vt) {
+    while (vt != jl_any_type) {
```
Is this sufficient stopping criteria to make sure we never infinite loop? Should I also check for NULL, for example?

Keno replied:
Yes, this is fine.

NHDaly replied:
thanks! :)

vtjnash replied:
because of where this gets called from, yeah, you should probably check for NULL also

Keno replied:
Yeah, I was just about to come back and say that. This function is regularly called on invalid data. You might also want to include a max trip count or something in case the datatype loops.

vtjnash replied:
It shouldn't be quite that invalid, but we might end up here to print out a message about the incomplete initialization of some struct type

Keno replied:
I usually like to be as defensive as possible in these functions, since I not-irregularly call them using `jl_` on complete garbage data, which could easily have cycles.

NHDaly replied:
What is a reasonable max number of iterations? 100? Do we have type hierarchies bigger than that? 1000?

Pointer chasing is pretty fast, so i don't seem much harm in a reasonably large number?

NHDaly replied:
Does it make sense to throw an exception if I encounter a null and/or infinite loop? Or should I just print a warning and then return false?

NHDaly replied:
Okay, I've pushed up 2479af7, which fixes this, by throwing an error after 10,000 iterations. It happens imperceptibly fast, so it seems fine to me:
```julia
julia> v = ()->2+2
#1 (generic function with 1 method)

julia> vt = typeof(v)
var"#1#2"

julia> vt.super = vt;

julia> x = v;

julia> p = Pipe()
Pipe(RawFD(4294967295) init => RawFD(4294967295) init, 0 bytes waiting)

julia> Base.link_pipe!(p, reader_supports_async=true, writer_supports_async=true)
Pipe(RawFD(21) open => RawFD(20) open, 0 bytes waiting)

julia> ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x)
ERROR: static_show: Exit after 10,000 iterations of supertype(). Presuming invalid cycle in datatype object.
Stacktrace:
 [1] top-level scope
   @ ./REPL[7]:1

julia> @time try; ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x) catch; end
  0.000047 seconds (3 allocations: 160 bytes)
```

---------

I think this should be ready for another look. Thanks both! :)

yuyichao replied:
Errors are handled differently in this function and in particular you shouldn't call `jl_error` since it allocates.
Just print a warning is fine.

NHDaly replied:
Oh, shoot make sense.

Okay, i'll switch to print a warning. Thanks @yuyichao!

Sorry, one more question:
- `jl_printf((JL_STREAM*)STDERR_FILENO,` or `jl_printf(JL_STDERR,`?
    - I see both examples in this file. I've gone with the first one, since it seems to be used in the `jl_static_*` functions?

Done (for now) in 7e2d96736a:
```julia
julia> ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x)
ERROR: static_show: Exit after 10,000 iterations of supertype(). Presuming invalid cycle in datatype object.
```

vtjnash replied:
FWIW, the rest of this code also used to use an iteration maximum, then print `•`, but it was removed in favor of explicit tracking with `struct recur_list *depth`. The helper functions, such as this one, `first_arg_datatype` and `jl_static_show_func_sig` have previously ignored those invalid possibilities.

NHDaly replied:
Hrmm those are good points...

@vtjnash - without the check for infinite loops, the above example _definitely does_ infinite loop:
```julia
julia> ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x)
^C^C^C^C^CWARNING: Force throwing a SIGINT
ERROR: ^C^C^C^C^C^CInterruptException:
Stacktrace:
 [1] top-level scope
   @ ./REPL[7]:1
```

Is there some way that I should use this `recur_list *depth`? I'm not exactly sure what you're recommending based on the above insight.

Based on the feedback from above, my current thinking is that I'll just remove the warnings, and return false from this function if we encounter a loop or a NULL, since both of those mean: it's not a function!

Does that sound good to you all?

NHDaly replied:
I've pushed up aa77e3528e which makes the above change. Please take another look if you can! :)

NHDaly committed:
PR feedback: JL_NOTSAFEPOINT, jl_function_type

NHDaly committed:
Add error checking to `jl_static_is_function_`: null and cycles

NHDaly committed:
Print warning to STDERR instead of throwing exception

Since throwing an exception would allocate, which is illegal from the
`jl_static_*` function(s).

```
julia> ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x)
ERROR: static_show: Exit after 10,000 iterations of supertype(). Presuming invalid cycle in datatype object.
```

Sacha0 reviewed (approved):
LGTM insofar as I can assess this! :)

yuyichao reviewed (commented):
......... I wrote this 3 hours ago and just noticed that I didn't submit the review = = ........
I'm still failing for the submit review vs submit single comment now = = ....

vtjnash commented on src/rtutils.c:
```diff
@@ -641,6 +641,32 @@ static size_t jl_static_show_x_sym_escaped(JL_STREAM *out, jl_sym_t *name) JL_NO
+    while (vt != jl_any_type) {
+        if (vt == NULL) {
+            jl_printf((JL_STREAM*)STDERR_FILENO,
+                "ERROR: static_show: Nullptr encountered inside datatype.\n");
```
This isn't even an error, just the appearance of the #undef value. We format it as `#<null>` usually.

NHDaly replied:
Thanks - removed this, per the above.

NHDaly committed:
Remove WARNING logs on `jl_static_is_function_()`

If we ecounter a cycle or a NULL super type before encountering a
Function super type, then this type is not <: Function.

NHDaly reviewed (commented):
Oops, i always forget to click re-request review.

Review re-requested. PTAL thanks!

NHDaly commented:
Thanks! :)

NHDaly commented:
If/when we do a 1.5.4 release, this should be included in it. Please let us know if we can help with that in any way!
